addon/Feeds/BookmarksFeed.js

```diff
@@ -0,0 +1,105 @@
+const {Cu} = require("chrome");
+const {PlacesProvider} = require("addon/PlacesProvider");
+const Feed = require("addon/lib/Feed");
+const {BOOKMARKS_LENGTH} = require("common/constants");
+const am = require("common/action-manager");
+const getScreenshot = require("addon/lib/getScreenshot");
+
+Cu.import("resource://gre/modules/Task.jsm");
+
+const UPDATE_TIME = 15 * 60 * 1000; // 15 minutes
+
+module.exports = class BookmarksFeed extends Feed {
+  constructor(options) {
+    super(options);
+    this.getScreenshot = getScreenshot;
+    this.missingData = false;
+  }
+
+  /**
+   * shouldGetScreenshot - Returns true if the link/site provided meets the following:
+   * - is a bookmark
+   * - has metadata
+   * - doesn't have any images
+   *
+   * @return bool
+   */
+  shouldGetScreenshot(link) {
+    return link.bookmarkGuid &&
+      link.hasMetadata &&
+      (!link.images || link.images.length === 0);
+  }
+
+  /**
+   * getData
+   *
+   * @return Promise  A promise that resolves with the "BOOKMARKS_RESPONSE" action
+   */
+  getData() {
+    return Task.spawn(function*() {
+      let links;
+      // Get links from places
+      links = yield PlacesProvider.links.getBookmarks({limit: BOOKMARKS_LENGTH});
```

Comment:
after playing with it locally, i think we should request some extras for the case that the user removes a bookmark from the context menu. that way the row stays filled in with 3 if some are removed.